clusters/resource_cluster.go

```diff
@@ -123,62 +123,62 @@ type ClusterSpec struct {
 }
 
-func (ClusterSpec) CustomizeSchema(s map[string]*schema.Schema) map[string]*schema.Schema {
-	common.CustomizeSchemaPath(s, "cluster_source").SetReadOnly()
-	common.CustomizeSchemaPath(s, "enable_elastic_disk").SetComputed()
-	common.CustomizeSchemaPath(s, "enable_local_disk_encryption").SetComputed()
-	common.CustomizeSchemaPath(s, "node_type_id").SetComputed().SetConflictsWith([]string{"driver_instance_pool_id", "instance_pool_id"})
-	common.CustomizeSchemaPath(s, "driver_node_type_id").SetComputed().SetConflictsWith([]string{"driver_instance_pool_id", "instance_pool_id"})
-	common.CustomizeSchemaPath(s, "driver_instance_pool_id").SetComputed().SetConflictsWith([]string{"driver_node_type_id", "node_type_id"})
-	common.CustomizeSchemaPath(s, "ssh_public_keys").SetMaxItems(10)
-	common.CustomizeSchemaPath(s, "init_scripts").SetMaxItems(10)
-	common.CustomizeSchemaPath(s, "init_scripts", "dbfs").SetDeprecated(DbfsDeprecationWarning)
-	common.CustomizeSchemaPath(s, "init_scripts", "dbfs", "destination").SetRequired()
-	common.CustomizeSchemaPath(s, "init_scripts", "s3", "destination").SetRequired()
-	common.CustomizeSchemaPath(s, "init_scripts", "volumes", "destination").SetRequired()
-	common.CustomizeSchemaPath(s, "init_scripts", "workspace", "destination").SetRequired()
-	common.CustomizeSchemaPath(s, "workload_type", "clients").SetRequired()
-	common.CustomizeSchemaPath(s, "workload_type", "clients", "notebooks").SetDefault(true)
-	common.CustomizeSchemaPath(s, "workload_type", "clients", "jobs").SetDefault(true)
-	s["library"].Set = func(i any) int {
+func (ClusterSpec) CustomizeSchema(s *common.CustomizableSchema) *common.CustomizableSchema {
+	s.SchemaPath("cluster_source").SetReadOnly()
+	s.SchemaPath("enable_elastic_disk").SetComputed()
+	s.SchemaPath("enable_local_disk_encryption").SetComputed()
+	s.SchemaPath("node_type_id").SetComputed().SetConflictsWith([]string{"driver_instance_pool_id", "instance_pool_id"})
+	s.SchemaPath("driver_node_type_id").SetComputed().SetConflictsWith([]string{"driver_instance_pool_id", "instance_pool_id"})
+	s.SchemaPath("driver_instance_pool_id").SetComputed().SetConflictsWith([]string{"driver_node_type_id", "node_type_id"})
+	s.SchemaPath("ssh_public_keys").SetMaxItems(10)
+	s.SchemaPath("init_scripts").SetMaxItems(10)
+	s.SchemaPath("init_scripts", "dbfs").SetDeprecated(DbfsDeprecationWarning)
+	s.SchemaPath("init_scripts", "dbfs", "destination").SetRequired()
+	s.SchemaPath("init_scripts", "s3", "destination").SetRequired()
+	s.SchemaPath("init_scripts", "volumes", "destination").SetRequired()
+	s.SchemaPath("init_scripts", "workspace", "destination").SetRequired()
+	s.SchemaPath("workload_type", "clients").SetRequired()
+	s.SchemaPath("workload_type", "clients", "notebooks").SetDefault(true)
+	s.SchemaPath("workload_type", "clients", "jobs").SetDefault(true)
+	s.SchemaPath("library").Schema.Set = func(i any) int {
```

Comment:
New interface would hide the details about `SchemaPathContext`.

And I think it is not very hard to use, might even be easier since we don't have to type `common.CustomizeSchemaPath` on each line.

Let me know what you think @mgyucht

Reply:
Meanwhile, only the new resources using `ResourceProvider` are onboarded to this change. The legacy ones will have their `CustomizeSchema` callback function unchanged because 
  1. there will be a lot of files we need to refactor if we change them 
  2. the problem that `SchemaPathContext` is trying to solve only applies to those using `ResourceProvider`, so it's not necessary
  3.  the legacy `CustomizeSchema` callback functions are using a lot of direct modifications of the `*map[string]*schema.Schema` object, which requires extra step to refactor into this new interface